scipy/integrate/_quadrature.py

```diff
@@ -482,24 +474,24 @@ def simpson(y, x=None, *, dx=1.0, axis=-1):
             slice2 = tupleset(slice_all, axis, -2)
             if x is not None:
                 last_dx = x[slice1] - x[slice2]
-            val += 0.5 * last_dx * (y[slice1] + y[slice2])
+            val = val + 0.5 * last_dx * (y[slice1] + y[slice2])
         else:
             # use Simpson's rule on first intervals
-            result = _basic_simpson(y, 0, N-3, x, dx, axis)
+            result = _basic_simpson(y, 0, N-3, x, dx, axis, xp=xp)
 
             slice1 = tupleset(slice_all, axis, -1)
             slice2 = tupleset(slice_all, axis, -2)
             slice3 = tupleset(slice_all, axis, -3)
 
-            h = np.asarray([dx, dx], dtype=np.float64)
+            h = xp.stack([dx, dx])
```

Comment:
Again, `dtype` here and below is taken care of by `xp_promote`. Yes, forcing `float64` is being fixed as a bug.